packages/flutter/lib/src/widgets/animated_child_switcher.dart

```diff
@@ -10,21 +10,64 @@ import 'framework.dart';
 import 'ticker_provider.dart';
 import 'transitions.dart';
 
+// Internal representation of a child that, now or in the past, was set on the
+// AnimatedChildSwitcher.child field, but is now in the process of
+// transitioning. The internal representation includes fields that we don't want
+// to expose to the public API (like the controller).
 class _AnimatedChildSwitcherChildEntry {
   _AnimatedChildSwitcherChildEntry(this.widget, this.controller, this.animation);
 
+  // The widget that is currently, or used to be, set as the widget's child.
   Widget widget;
 
+  // The animation controller for the child's transition.
   final AnimationController controller;
+
+  // The animation describing the child's transition, with the controller as
+  // its parent.
+  final Animation<double> animation;
+}
+
+/// Represents a child that used to be set on the [AnimatedChildSwitcher.child]
+/// field, but is now in the process of transitioning out.
+class AnimatedChildSwitcherLayoutData {
+  AnimatedChildSwitcherLayoutData(this.widget, this.animation);
+
+  /// The widget that is currently or used to be set as the widget's child.
+  Widget widget;
+
+  /// The animation describing the child's transition. It is transitioning in
+  /// if the animation is running forwards.
   final Animation<double> animation;
 }
 
-/// A widget that automatically does a [FadeTransition] between a new widget and
+/// The type of the builder used to generate custom transitions for
+/// [AnimatedChildSwitcher].
+///
+/// The [child] should be transitioned in when [animation] is running in the
+/// forward direction,
+typedef Widget AnimatedChildSwitcherTransitionBuilder(Widget child, Animation<double> animation);
+
+/// The type of the builder used to generate custom layouts for
+/// [AnimatedChildSwitcher].
+///
+/// The [children] should be positioned according to their corresponding animation.
+typedef Widget AnimatedChildSwitcherLayoutBuilder(List<AnimatedChildSwitcherLayoutData> children);
+
+/// A widget that by default does a [FadeTransition] between a new widget and
 /// the widget previously set on the [AnimatedChildSwitcher] as a child.
 ///
-/// More than one previous child can exist and be fading out while the newest
-/// one is fading in if they are swapped fast enough (i.e. before [duration]
-/// elapses).
+/// If they are swapped fast enough (i.e. before [duration] elapses), more than
+/// one previous child can exist and be transitioning out while the newest one
+/// is transitioning in.
+///
+/// If the "new" child is the same widget type as the "old" child, but with
+/// different parameters, then [AnimatedChildSwitcher] will *not* do a
+/// transition between them, since as far as the framework is concerned, they
+/// are essentially the same widget, and the existing widget can be updated with
```

Comment:
strike "essentially"

Reply:
though then this reads a bit weird, since you are sometimes saying they're the same and sometimes saying they're different.